technical.js

```diff
@@ -1,221 +1,261 @@
-},
+  TR(x0, x) {
+    return Math.max(
+      x.high - x.low,
+      Math.abs(x.high - x0.close),
+      Math.abs(x.low - x0.close)
+    );
+  },
 
-ATR: function(x, length) {
-    for (let i = x.length-length-2; i >= 0; i--) {
-        var prev_atr = x[i+1]['ATR'];
-        if (typeof prev_atr === 'undefined' || isNaN(prev_atr)) {
-            let sum = 0;
-            for (j = i+1; j < x.length-1; j++) {
-                sum += this.TR(x[j+1], x[j]);
-            }
-            prev_atr = sum / length;
+  ATR(x, length) {
+    for (let i = x.length - length - 2; i >= 0; i--) {
+      let prev_atr = x[i + 1]["ATR"];
+      if (typeof prev_atr === "undefined" || isNaN(prev_atr)) {
+        let sum = 0;
+        for (j = i + 1; j < x.length - 1; j++) {
+          sum += this.TR(x[j + 1], x[j]);
         }
-        x[i]['ATR'] = (prev_atr * (length-1) + this.TR(x[i+1], x[i])) / length
+        prev_atr = sum / length;
+      }
+      x[i]["ATR"] =
+        (prev_atr * (length - 1) + this.TR(x[i + 1], x[i])) / length;
     }
-    return x[0]['ATR'];
-},
+    return x[0]["ATR"];
+  },
 
-MACDH: function(x, offset, source_key) {
+  MACDH(x, offset, source_key) {
     if (!offset) offset = 0;
-    if (!source_key) source_key = 'close';
-    this.EMA(x, 12, 'ema_short', source_key);
-    this.EMA(x, 26, 'ema_long', source_key);
-    x.forEach(function(period) {
-        if (period['ema_short'] !== undefined && period['ema_long'] !== undefined) 
-            period['macd'] = period['ema_short'] - period['ema_long'];
+    if (!source_key) source_key = "close";
+    this.EMA(x, 12, "ema_short", source_key);
+    this.EMA(x, 26, "ema_long", source_key);
+    x.forEach(period => {
+      if (period["ema_short"] !== undefined && period["ema_long"] !== undefined)
+        period["macd"] = period["ema_short"] - period["ema_long"];
     });
-    this.EMA(x, 9, 'signal', 'macd');
-    return (x[offset]['macd'] - x[offset]['signal']);
-},
+    this.EMA(x, 9, "signal", "macd");
+    return x[offset]["macd"] - x[offset]["signal"];
+  },
 
-SAR: function(x) {
-    let iaf = 0.02, maxaf = 0.2;
+  SAR(x) {
+    let iaf = 0.02;
+    let maxaf = 0.2;
     let bull = true;
     let af = iaf;
-    let lp = x[x.length-1].low;
-    let hp = x[x.length-1].high;
-    for (let i = x.length-3; i >= 0; i--) {
-        if (x[i+1].psar === undefined)
-            x[i+1].psar = x[i+1].close;
-        if (bull)
-            x[i].psar = x[i+1].psar + af * (hp - x[i+1].psar)
-        else
-            x[i].psar = x[i+1].psar + af * (lp - x[i+1].psar)
-        var reverse = false;
+    let lp = x[x.length - 1].low;
+    let hp = x[x.length - 1].high;
+    for (let i = x.length - 3; i >= 0; i--) {
+      if (x[i + 1].psar === undefined) x[i + 1].psar = x[i + 1].close;
+      if (bull) x[i].psar = x[i + 1].psar + af * (hp - x[i + 1].psar);
+      else x[i].psar = x[i + 1].psar + af * (lp - x[i + 1].psar);
+      let reverse = false;
```

Comment:
reverse must be function scope (var) or defined above.

Reply:
Ah that's fair. II'll give this a slower pass sometime this weekend.